Optimize page loading by lazy loading backend configuration and pipeline data

Karakatiza666 opened the pull request:
This PR dramatically improves UX on slower connections - first page load is faster, navigation between pages is instant. It also contains other quality-of-life fixes.
This PR combines many changes because most of them affect significant surface area, prompting extensive testing; also many changes are interconnected. Manual testing of changes in sequence, or distilling and rebasing into a sequence of PRs would take a long time, while the total time of code review would not decrease.

**Testing:  extensively manually tested** most UI features as well as behavior with authentication enabled, when killing the API server, and when changing authentication configuration mid-session.

- Configuration caching
Now the backend configuration (version, telemetry etc.) is cached to render the page initially. Once the page is rendered, the configuration is lazy loaded and related values are updated.

- Refactor pipeline data loading system - lazy loading of pipelines data and demos
This makes the web-console responsive and switching between the pages is instant even on slow connections.
While pipeline list, pipeline code or demos are loading the placeholders are shown:
<img width="2673" height="1649" alt="image" src="https://github.com/user-attachments/assets/f941d002-7e1c-4715-a42e-aa6a0625e71e" />

<img width="2673" height="1649" alt="image" src="https://github.com/user-attachments/assets/1a5e4b64-4ac6-4a9a-8785-d01ad01a6b14" />

<img width="2670" height="1787" alt="image" src="https://github.com/user-attachments/assets/c3449e0b-1257-4a18-8cab-8f31059aa33a" />
<img width="2670" height="1787" alt="image" src="https://github.com/user-attachments/assets/d78a4a21-4e42-46d3-b217-185c62ad3f10" />

- Monolith bundle
Usually it is recommended to split the web app code in chunks that are loaded as needed. In our case when the customer has a very slow connection / pipeline-manager is bottle-necked, it is faster to load a single large file rather than multiple smaller ones. If the web app is cached properly the bundle size becomes irrelevant in subsequent requests.

- Prevent leaving pipeline with unsaved changes
Modal prompt to save or discard unsaved pipeline changes before switching pipeline or leaving the page
<img width="3082" height="1922" alt="image" src="https://github.com/user-attachments/assets/f791e60b-5555-4f87-ae1f-7f8d99c3ef2f" />

<img width="2576" height="1098" alt="image" src="https://github.com/user-attachments/assets/f636bd09-63e9-4aa1-8684-17f3f7a2175a" />

Update 24.04.26:

Improved:
- Gracefully display a popup prompting to re-login on unrecoverable authentication error (e.g. when authentication config changed, or secret rotated mid-session)
<img width="4481" height="726" alt="image" src="https://github.com/user-attachments/assets/e4dcf0ac-6a6a-4aa8-b63c-15128db54923" />

- Update openapi-ts version and re-generate API binding files. 
Actualize the dependency that is responsible for generating javascript code to query endpoints defined in OpenAPI spec. Also: change codegen settings for better devX
- Disable buttons to download pipeline profile when already loading/uploading a profile

Fixed:
- Duplicate HTTP requests for cluster health events

Technical:
- Better separation of concerns of Pipeline Code Editor implementation

[Screencast from 2026-04-23 14-40-23.webm](https://github.com/user-attachments/assets/90954a69-8959-4caf-88fc-67a27657d652)

gz reviewed (approved):
looks like a big change and the worry is things don't work anymore. make sure to test it well...

Karakatiza666 commented:
The current plan is to merge it at least a week later, in the meantime I will do some more testing. I already discovered some corner case issue that happens only on a slow connection.

mythical-fred reviewed (changes requested):
Large behavioral change — lazy loading, config caching, new localStorage compositions — with no tests.

Karakatiza666 committed:
[web-console] Change data loading approach from pre-loading to lazy loading

Reduce coupling by distilling PipelineCodePanel.svelte from PipelineEditLayout.svelte and CodeEditor.svelte

Signed-off-by: Karakatiza666 <bulakh.96@gmail.com>

Karakatiza666 committed:
[web-console] Refactor API client generation

Improve handling network errors + other fixes

Signed-off-by: Karakatiza666 <bulakh.96@gmail.com>

Karakatiza666 commented on js-packages/web-console/src/lib/compositions/useLocalStorage.svelte.ts:
```diff
@@ -0,0 +1,251 @@
+/**
```
Ended up removing this new implementation

mihaibudiu commented on js-packages/web-console/src/lib/compositions/localStore.svelte.js:
```diff
@@ -2,7 +2,8 @@
 export class LocalStore {
-  value = $state(/** @type {T} */ (/** @type {unknown} */ undefined))
+  /** @type {T} */
+  #value
```
Why not `#value: T` and no `T` as type parameter?
In general, why is T only in comments in this file?

Karakatiza666 replied:
This is a plain JS file, so it uses special JS Doc syntax for type annotations in JS comments that still get recognized by the syntax highlight tools and the typescript compiler; the origin of the file was a snippet from the Internet which I did not convert to TypeScript. I should convert it to TypeScript at some point, but there's no pressure for it.

mihaibudiu commented on js-packages/web-console/src/lib/compositions/pipelines/useCodeEditorSettings.svelte.ts:
```diff
@@ -1,4 +1,5 @@
-import { useLocalStorage } from '$lib/compositions/localStore.svelte'
+import { useLocalStorage } from '$lib/compositions/localStore.svelte.js'
+// import { useLocalStorage } from '$lib/compositions/useLocalStorage.svelte'
```
should this be deleted?

mihaibudiu commented on js-packages/web-console/src/lib/components/pipelines/list/Actions.svelte:
```diff
@@ -580,7 +592,7 @@ groups related actions into multi-action dropdowns when multiple options are ava
               <span class="">{button.label}</span>
 
               {#if button.disabled}
```
This seems to have changed behavior

Karakatiza666 replied:
"Enterprise Only" is now not the only reason to prevent clicking an action, so I am now passing the short tip as an argument instead of hard-coding

mihaibudiu commented on js-packages/web-console/src/lib/components/pipelines/editor/CodeEditor.svelte:
```diff
@@ -102,17 +95,69 @@
+  }
+
+  /**
+   * Reveal a specific line and column in the editor for a given file.
```
should this read "for the current file?"
Then you need to delete the fileName param

mihaibudiu commented on js-packages/web-console/src/lib/components/pipelines/editor/CodeEditor.svelte:
```diff
@@ -102,17 +95,69 @@
+      }))
+    )
+    // Reveal the first selection in center
+    editorRef.revealRangeInCenter({
```
whose center? I suspect it's the screen

mihaibudiu commented on js-packages/web-console/src/lib/components/pipelines/editor/CodeEditor.svelte:
```diff
@@ -390,6 +328,12 @@
       Please resolve the conflict to save your changes.
     </div>
     <div class="flex flex-nowrap justify-end gap-4">
+      <button
```
what is this?

Karakatiza666 replied:
You may not have seen this feature before - it's a popup on the bottom-right of the editor for cases that the pipeline code was changed remotely while you had unsaved changes. The new option allows you to keep editing without overwriting remote with your local changes, which was not available before.

mihaibudiu commented on js-packages/web-console/src/lib/components/pipelines/List.svelte:
```diff
@@ -14,7 +14,7 @@
   }: {
     pipelineName: string
-    pipelines: PipelineThumb[]
+    pipelines: PipelineThumb[] | undefined
```
Technically an empty array could be the same as undefined. But this is fine too

Karakatiza666 replied:
I strictly need to know the difference between pipeline list being empty and being loaded

mihaibudiu commented on js-packages/web-console/src/lib/compositions/usePipelineManager.svelte.ts:
```diff
@@ -86,16 +91,31 @@ const _reportError =
 
   const doNotReportIfCancelled = (error: Error) => {
-    return error.cause === 'cancelled'
+    // Explicit user-triggered cancellation (pipelineManager streaming fetches
```
can you document the meaning of the return value?

mihaibudiu commented on js-packages/web-console/src/lib/services/auth.ts:
```diff
@@ -85,6 +85,63 @@ export const authResponseMiddleware = async (response: Response, request: Reques
+export const errorResponseMiddleware = (error: unknown, response: Response | undefined) => {
+  if (error && typeof error === 'object') {
+    try {
+      ;(error as any).response = response
```
why is the semicolon needed here?

Karakatiza666 replied:
This is a defensive choice by a formatter to avoid the parentheses to be treated as a function call arguments if a line is added above this one. Not required here.

mihaibudiu commented on js-packages/web-console/src/lib/components/layout/pipelines/UnsavedPipelineChanges.svelte:
```diff
@@ -0,0 +1,136 @@
+    if (!unsavedChanges || navigationResolve || suppressNavigationGuard) {
+      return
+    }
+    const samePage = (() => {
```
why define a function if you call it right away?
why not just inline the function body?

Karakatiza666 replied:
Inlining the function body would spill the temporary variables - `fromUrl` and `toUrl` - to the scope of the rest of the parent function. I often use IIFEs (Immediately Invoked Function Expression) for this to make code cleaner (IMO) - when re-visiting the code I see immediately that a certain variable will not be needed after a block I am not currently interested in, so I can mentally "garbage collect" it.

mihaibudiu commented on js-packages/web-console/src/lib/components/layout/pipelines/PipelineEditLayout.svelte:
```diff
@@ -1,233 +1,92 @@
+    pipelineName: string
+    pipelineList: { pipelines: PipelineThumb[] | undefined } // If it's undefined - it is loading
+    pipelineThumb: PipelineThumb | undefined // If it's undefined - it is loading
+    pipeline: WritablePipeline // If it's undefined - it is loading
```
there's no `| undefined` here

Karakatiza666 replied:
Will clarify the comment

mihaibudiu commented on js-packages/web-console/src/lib/components/layout/pipelines/PipelineEditLayout.svelte:
```diff
@@ -485,136 +334,112 @@ example = "1.0"`
+{/snippet}
+{#snippet statusBarEnd()}
+  <div class="ml-auto flex flex-nowrap items-center gap-1">
+    {#each [{ icon: 'fd fd-panel-left', text: 'Pipelines', value: showPipelinesPanel }, { icon: 'fd fd-panel-bottom', text: 'Monitoring', value: showMonitoringPanel }, { icon: 'fd fd-panel-right', text: 'Interaction', value: showInteractionPanel, show: isScreenLg.current }] as { icon, text, value, show }}
```
what a long line

Karakatiza666 replied:
Will move to a named variable so the values are formatted nicely, multiline

mihaibudiu commented on js-packages/web-console/src/routes/+layout.ts:
```diff
@@ -28,6 +37,50 @@ const { OidcClient, OidcLocation } = AxaOidc
+ *     does not re-enter `lazyUpdateConfig()`.
+ *   - Unauthenticated: auth resolves to `{ login }`  →  empty layout is
+ *     returned; `(authenticated)/+layout.ts` triggers `auth.login()` and the
+ *     app redirects to the IdP.
```
IdP?

mihaibudiu commented on js-packages/web-console/src/routes/+layout.ts:
```diff
@@ -70,58 +123,246 @@ const emptyLayoutData: LayoutData = {
+ * Hardcoded toggle for the warm-cache optimistic render. When `false`, every
+ * `load()` blocks on `fetchConfigs()` (same shape as the cold-cache path) and
+ * the `lazyUpdateConfig()` reconcile is skipped. The cache itself is still
+ * written by `fetchConfigs()`, so flipping this back to `true` resumes
```
back?

Karakatiza666 replied:
This reads fine to me, but I'll rephrase this part to avoid ambiguity

mihaibudiu commented on js-packages/web-console/src/routes/+layout.ts:
```diff
@@ -70,58 +123,246 @@ const emptyLayoutData: LayoutData = {
+ *
+ * Server time is intentionally NOT restored from cache: the cached
+ * `license.current` timestamp is stale and applying it would corrupt the
+ * offset. Until fresh data arrives we let `Date.now()` stand in.
```
offset?

Karakatiza666 replied:
This is an implementation detail, revised the comment

mihaibudiu commented on js-packages/web-console/src/lib/components/layout/pipelines/PipelineCodePanel.svelte:
```diff
@@ -0,0 +1,401 @@
+    const cbs = pipelineActionCallbacks
+      .getAll('', 'delete')
+      .concat(pipelineActionCallbacks.getAll(pipelineName, 'delete'))
+    cbs.map((x) => x(pipelineName))
```
rename x to cb

mihaibudiu reviewed (approved):
Can't say that I understood everything, but I checked superficially

mythical-fred reviewed (changes requested):
Re-checking after the force-push — my prior REQUEST_CHANGES still stands. The diff still touches ~50 web-console files with deep behavioral changes (lazy loading, config caching, localStorage compositions, auth error handling, layout reorgs), and the only test changes are 6 lines in `DeletedPipelineState.svelte.test.ts` and 3 in `promise.svelte.spec.ts`. None of the new behavior — `configCache`, `useDemos`, `UnsavedPipelineChanges`, the rewritten `+layout.ts` boot sequence, the `usePipelineManager` rework — has any test coverage.

Per the standing rule for web-console PRs, behavior changes without tests are a hard block, even with approvals from other reviewers. Vitest + `@testing-library/svelte` is already wired up; please add coverage for at least the cache hit/miss/invalidation paths and the lazy-load failure path before this is mergeable.

Karakatiza666 committed:
[web-console] Lazy loading PR review fixes, add unit tests for config caching

Signed-off-by: Karakatiza666 <bulakh.96@gmail.com>
